### PR TITLE
Plate set perf improvements

### LINKED ISSUE #5696: AbDisc: Perf regression investigation
**Tasks**
- [x] Review failures : 2
- [x] Analyze code changes : 2
- [x] Profile plate set creation : 2
- [x] Defer container delete in `PlateSetPerfTest` : 1
- [x] Review new baseline : 3
- [x] Review table indices : 3
- [x] Defer plate indexing in bulk operations : 3

### PR DESCRIPTION
**Pull request by labkey-klum**
#### Rationale
- Expose some new `PlateSetDataGenerator` metrics:
  - Time to create per plate.
  - Std deviation of per plate creation times. This isn't exactly right since for each plate set we are taking the average based on plate set creation time versus number of plates. We currently can't break it down further for multiple plates in a plate set.
- Adds an index for the plate metadata table `lsid` field.
- For plate set creation, delay the indexing of plates in the plate set. Once the plate set has been created, a background task is run to index each individual plate. 

#### Related Pull Requests
https://github.com/LabKey/limsModules/pull/484

### CODE REVIEW

File: assay/src/org/labkey/assay/plate/PlateManager.java [hunk 1]
```diff
@@ -2397,6 +2433,7 @@ private List<Plate> addPlatesToPlateSet(
                 platesAdded.add(createAndSavePlate(container, user, plateImpl, plateSetId, plate.data));
             }
 
+            tx.addCommitTask(this::resumePlateIndexing, DbScope.CommitTaskOption.POSTCOMMIT);
```
Comment (labkey-nicka) [inline assay/src/org/labkey/assay/plate/PlateManager.java:2436]: I think we need to do something to resume plate indexing in the event this commit fails.

```java
tx.addCommitTask(this::resumePlateIndexing, DbScope.CommitTaskOption.POSTCOMMIT, DbScope.CommitTaskOption.POSTROLLBACK);
```

File: assay/src/org/labkey/assay/plate/PlateMetadataDomainKind.java [hunk 2]
```diff
@@ -24,34 +25,45 @@
...
 
 public class PlateMetadataDomainKind extends BaseAbstractDomainKind
 {
     public static final String KIND_NAME = "PlateMetadata";
     public static final String DOMAiN_NAME = "PlateMetadataDomain";
     private static final Set<String> RESERVED_NAMES;
+    private static final Set<PropertyStorageSpec.Index> INDEXES;
     private static final String PROVISIONED_SCHEMA_NAME = "assaywell";
 
+    public enum Column
```
Comment (labkey-nicka) [inline assay/src/org/labkey/assay/plate/PlateMetadataDomainKind.java]: Embrace the enum! 👏

File: assay/src/org/labkey/assay/plate/PlateManager.java [hunk 3]
```diff
@@ -176,6 +179,10 @@ public class PlateManager implements PlateService, AssayListener, ExperimentList
     private static final String PLATE_SET_NAME_EXPRESSION = "PLS-${now:date('yyyyMMdd')}-${RowId}";
     private static final String PLATE_NAME_EXPRESSION = "${${PlateSet/PlateSetId}-:withCounter}";
 
+    private final Queue<Pair<Container, Integer>> _plateIndexQueue = new ConcurrentLinkedQueue<>();
```
Comment (labkey-nicka) [inline assay/src/org/labkey/assay/plate/PlateManager.java]: Consider using a `Queue<Map<Container, List<Integer>>>` to reduce number of container instances held in memory.

File: assay/src/org/labkey/assay/plate/PlateManager.java [hunk 4]
```diff
@@ -2385,6 +2420,7 @@ private List<Plate> addPlatesToPlateSet(
 
         try (DbScope.Transaction tx = ensureTransaction())
         {
+            pausePlateIndexing();
```
Comment (labkey-nicka) [inline assay/src/org/labkey/assay/plate/PlateManager.java:2423]: Might be worth considering somehow pinning this to a plate set as we're operating solely within a plate set scope here. Seems like we could get into race conditions where the queue gets resumed while someone else is in the middle of creating plates on a different plate set.

Possibly could make use of a concurrent set containing plate set rowIds or something like that.

File: assay/src/org/labkey/assay/plate/PlateMetadataDomainKind.java [hunk 5]
```diff
@@ -68,7 +80,7 @@ public static String generateDomainURI(@NotNull Container container)
     @Override
     public Set<PropertyStorageSpec> getBaseProperties(Domain domain)
     {
-        return Collections.singleton(new PropertyStorageSpec("Lsid", JdbcType.VARCHAR, 200).setNullable(false));
+        return Collections.singleton(new PropertyStorageSpec(Column.Lsid.name(), JdbcType.VARCHAR, 200).setNullable(false));
```
Comment (labkey-nicka) [inline assay/src/org/labkey/assay/plate/PlateMetadataDomainKind.java]: Just curious, did that have a noticeable impact on any queries/operations you saw?

File: assay/src/org/labkey/assay/plate/PlateManager.java (hunk 1 above)
Comment (labkey-klum) [reply assay/src/org/labkey/assay/plate/PlateManager.java:2436]: yep, done

File: assay/src/org/labkey/assay/plate/PlateManager.java (hunk 3 above)
Comment (labkey-klum) [reply assay/src/org/labkey/assay/plate/PlateManager.java]: done

**Review by labkey-alan — APPROVED** [review]
Looks good to me

**Pull request merged**